explcheck/src/prune-explcheck-config.lua

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env texlua
+-- A checker that reads the default configuration of the static analyzer explcheck and regression test results
+-- and then tests which parts of the configuration can be removed without affecting the results of the static analysis.
+
+local config = require("explcheck-config")
+local new_issues = require("explcheck-issues")
+local process_with_all_steps = require("explcheck-utils").process_with_all_steps
+
+local default_config = config.default_config
+local default_config_pathname = config.default_config_pathname
+local get_package = config.get_package
+
+-- Initialize and return a handle for KPathSea.
+local function initialize_kpse()
+  local kpse = require("kpse")
+  kpse.set_program_name("texlua", "explcheck")
+  return kpse
+end
+
+-- Read regression test results and use KPathSea to find the files in the results.
+local function read_results(results_pathname, kpse)
+  local num_skipped = 0
+  local results = {
+    filenames = {},
+    pathnames = {},
+    issues = {},
+  }
+  for line in io.lines(results_pathname) do
+    local line_filename, issues = line:match("^(%S+)%s+(%S+)$")
+    assert(line_filename ~= nil)
+    assert(issues ~= nil)
+    local line_pathname = kpse.find_file(line_filename)
+    if line_pathname == nil then
+      print(string.format('Could not determine the pathname of "%s" with KPathSea, skipping it.', line_filename))
+      num_skipped = num_skipped + 1
+      goto continue
+    end
+    table.insert(results.filenames, line_filename)
+    results.pathnames[line_filename] = line_pathname
+    results.issues[line_filename] = new_issues()
+    for issue in issues:gmatch("[^,]+") do
+      results.issues[line_filename]:add(issue)
+    end
+    ::continue::
+  end
+  if num_skipped > 0 then
+    print()
+  end
+  return results
+end
+
+-- For each file from regression test results, try to remove all options in the default configuration that apply to it and check
+-- whether this affects the results of the static analysis or not.
+local function main(results_pathname)
+  -- Read the results.
+  local kpse = initialize_kpse()
+  local results = read_results(results_pathname, kpse)
+
+  local num_filenames, num_options, num_possible_removals = 0, 0, 0
+  local seen_key_locations = {}
+
+  -- Try to remove a single option.
+  local function try_to_remove_option(pathname, key, key_location, default_value, expected_issues)
+    if seen_key_locations[key_location] ~= nil then
+      return
+    end
+    seen_key_locations[key_location] = true
```

Comment:
This doesn't work, since multiple pathnames can match the same `key_location`. Instead, we should collect results for all pathnames and then report only those `key_location`s for which the expected results match the actual results for all pathnames.

Reply:
Fixed in commit 654976c from this PR.